File: app/src/main/java/com/example/myapplication/ui/feed/FeedPresenter.java
package com.example.myapplication.ui.feed;

import androidx.annotation.NonNull;
import com.example.myapplication.data.Project;
import com.example.myapplication.usecases.ProjectsUseCase;
import com.example.myapplication.ui.BasePresenter;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

import javax.inject.Inject;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class FeedPresenter extends BasePresenter<FeedPresenter.View> {
    private final ProjectsUseCase projectsUseCase;

    public List<Project> projectList;

    @Inject
    public FeedPresenter(ProjectsUseCase projectsUseCase) {
        this.projectsUseCase = projectsUseCase;
    }


    private Disposable subscribeToLoadProjects(View view) {
        if (projectList == null) {
            return projectsUseCase.getProjects()
                    .doOnSubscribe(disposable -> view.showProgress())
                    .flatMapCompletable(projectList -> {
                        this.projectList = projectList;
                        if (projectList.isEmpty()) {
                            return view.showEmpty();
                        } else {
                            return view.showProjects(projectList);
                        }
                    })
                    .andThen(view.projectClick())
                    .doOnNext(view::goToProject)
                    .subscribe();
        } else {
            return view.showProjects(projectList)
                    .subscribe();
        }
    }

    @NonNull
    @Override
    protected Disposable onBind(View view) {
        return subscribeToLoadProjects(view);
    }

    public interface View extends BasePresenter.View {
        void showProgress();

        Completable showProjects(List<Project> projectList);

        Observable<Project> projectClick();

        Completable showEmpty();

        void goToProject(Project project);
    }
}

File: app/src/main/java/com/example/myapplication/ui/feed/FeedPresenter.kt
package com.example.myapplication.ui.feed

import android.annotation.SuppressLint
import com.example.myapplication.data.Projects
import com.example.myapplication.data.ProjectsUseCase
import com.example.myapplication.ui.BaseView
import com.hannesdorfmann.mosby3.mvi.MviBasePresenter
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.reactivex.subjects.BehaviorSubject
import javax.inject.Inject


//TODO A production app would also use a reducer
class FeedPresenter @Inject constructor(
	private val useCase: ProjectsUseCase
) : MviBasePresenter<FeedView, FeedViewState>() {
	@SuppressLint("CheckResult")
	override fun bindIntents() {
		val projects: Observable<FeedViewState> = intent(FeedView::loadProjects)
			.subscribeOn(Schedulers.io())
			.switchMap { useCase.getProjects() }
			.observeOn(AndroidSchedulers.mainThread())

		subscribeViewState(projects, FeedView::render)
	}
}

interface FeedView : BaseView<FeedViewState> {
	fun loadProjects(): BehaviorSubject<Unit>
}


sealed class FeedViewState {
	object Loading : FeedViewState()
	data class ProjectList(val projectList: List<Projects.Project>) : FeedViewState()
	data class Error(val msg: String) : FeedViewState()
}